elements/dash-tree/dash-tree.js

```diff
@@ -0,0 +1,448 @@
+ *
+ * Unlike the `DashRadioGroup`, which uses roving tabindex to indicate which
+ * child is currently active, the `DashTree` uses `aria-activedescendant` and
+ * the `id` of the currently active child. The effect is similar to using roving
+ * tabindex, and is presented in this case to show an alternative approach to
+ * indicating active children.
+ */
+
+(function() {
+  /**
+   * Define keycodes to help with handling keyboard events.
+   */
+  const KEYCODE = {
+    UP: 38,
+    DOWN: 40,
+    LEFT: 37,
+    RIGHT: 39,
+    HOME: 36,
+    END: 35,
+    SPACE: 32,
+    ENTER: 13,
+  };
+
+  /**
+   * A helper to quickly identify `DashTreeItem` nodes
+   */
+  function isTreeItem(node) {
+    return node.nodeName.toLowerCase() === 'dash-treeitem';
+  }
+
+  /**
+   * A helper to quickly identify ARIA states.
+   * Note that ARIA attributes require a literal string value of "true"/"false".
+   */
+  function isExpanded(node) {
+    return node.getAttribute('aria-expanded') === 'true';
+  }
+
+  // `dashTreeItemCounter` counts the number of `<dash-treeitem>` instances
+  // created. The number is used to generated new, unique IDs.
+  let dashTreeItemCounter = 0;
+
+  class DashTreeItem extends HTMLElement {
+    constructor() {
+      super();
+    }
+
+    connectedCallback() {
+      this.setAttribute('role', 'treeitem');
+
+      // If the element doesn't already have an `id`, generate one for it.
+      // This will make it easier to set the element as active using
+      // `aria-activedescendant`.
+      if (!this.id)
+        this.id = `dash-treeitem-generated-${dashTreeItemCounter++}`;
+
+      // If the element contains a `<dash-treegroup>` then it's a parent node
+      if (this.querySelector('dash-treegroup')) {
+        // If the element is not explicitly already expanded by the user, then
+        // set it to closed.
+        if (!isExpanded(this))
+          this.setAttribute('aria-expanded', false);
+
+        // Set the `aria-label` to the first bit of text. Otherwise the label
+        // for this element will be computed based on its text content plus
+        // the text content of all of its children. Making it so verbose as to
+        // be unusable. This first child should ideally be a `<span>` element,
+        // but it may also just be a line of text.
+        if (!this.hasAttribute('aria-label')) {
+          let label = this.querySelector('span');
+          if (!label) {
+            console.error(`The first child of a <dash-treeitem> that ` +
+              `contains a <dash-treegroup> must be a <span> containing label ` +
+              `text.`);
+          } else {
+            this.setAttribute('aria-label', label.textContent.trim());
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Define a custom element, `<dash-treeitem>`, and associate it with the
+   * `DashTreeItem` class.
+   */
+  window.customElements.define('dash-treeitem', DashTreeItem);
```

Comment:
Don’t punch me, but for consistency’s sake, shouldn’t it be `dash-tree-item` and `dash-tree-group`?